enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,275 @@
+- Respond to the attach/detach events of the disks/devices.
+- Have options for filtering particular kind of disks based on properties such as name, size, manufacturer, etc.
+
+### Non-Goals
+
+- Inherently protect against confilict with provisioners that own local devices on the same nodes automatic detection is configured to run.
+
+## Proposal
+
+The `local-storage-operator` is already capable of consuming local disks from the nodes and provisioning PVs out of them,
+but the disk/device paths needs to explicitly specified in the `LocalVolume` CR.
+The idea is to introduce a new Custom Resource, `AutoDetectVolumes`, that enables automated discovery and provisioning of local volume based PVs on a set of nodes addressed by one or more labelSelectors.
+
+### Risks and Mitigations
+
+- LSO will detect disks that contain data and/or are in-use via ensuring that the device:
+  - can be openened exclusively.
+  - is not read-only.
+  - is not removable.
+  - has no child partitions.
+  - has no FS signature.
+  - state (as outputted by `lsblk`) is not `suspended`
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+  - This is already ensured by the current LSO approach of consuming disks by their `UUID`
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```go
+
+type AutoDetectVolumeList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata"`
+    Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+    Spec              AutoDetectVolumeSpec   `json:"spec"`
+    Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+    // Nodes on which the automatic detection policies must run.
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+    // StorageClassName to use for set of matched devices
+    StorageClassName string `json:"storageClassName"`
+    // MinDeviceCount is the minumum number of devices that needs to be detected per node.
+    // If the match devices are less than the minCount specified then no PVs will be created.
+    MinDeviceCount int `json:"minDeviceCount"`
+    // +optional
+    // Maximum number of Devices that needs to be detected per node.
+    MaxDeviceCount int `json:"maxDeviceCount"`
+    // VolumeMode determines whether the PV created is Block or Filesystem
+    // + optional
+    VolumeMode PersistentVolumeMode `json:"volumeMode,omitempty"`
+    // FSType type to create when volumeMode is Filesystem
+    // +optional
+    FSType string `json:"fsType,omitempty"`
+    // DeviceInclusionSpec is the filtration rule for including a device in the device discovery
+    // +optional
+    DeviceInclusionSpec *DeviceInclusionSpec `json:"deviceInclusionSpec"`
+}
+
+type AutoDetectVolumeStatus struct {
+    Status string `json:"status"`
```

Comment:
Please expand this type.

Reply:
How about adding [pod conditions](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) instead of a single status string?

Reply:
Not sure adding pod conditions will be a good idea here. I think the status should be reflecting the discovery progress.

Reply:
I think expanding this, whatever you had in mind, would be overkill for an MVP. We should be able ot get by with just a simple string for the time being, unless you have something else specific in mind.

Reply:
what will be the support status of this new feature? I proposed to keep this alpha for now. but even then keeping this field as a simple string is not going to be enough. In LSO we typically try to stick to specs provided to us by api team - https://github.com/openshift/api/blob/master/operator/v1/types.go#L97

I am not sure if all fields in that type will be necessary but even for MVP/alpha feature, we should pick and propose fields we are going to use, so that this can be discussed during the design phase.